File: yunzhan.js
// ==UserScript==
// @name         云展网/yunzhan365 一键下载器 (v3.0 通用版)
// @namespace    http://tampermonkey.net/
// @version      3.0
// @license      MIT
// @description  修复 book.yunzhan365.com 无法下载的问题。支持所有云展网子域名，自动探测 files/large, files/mobile 等多种路径，自动去参，支持 PDF/ZIP 双模式。
// @author       Gemini Assistant
// @match        *://*.yunzhan365.com/*
// @match        *://*.fliphtml5.com/*
// @match        *://book.yunzhan365.com/*
// @match        *://bookh.yunzhan365.com/*
// @match        *://online.fliphtml5.com/*
// @require      https://cdnjs.cloudflare.com/ajax/libs/jszip/3.10.1/jszip.min.js
// @grant        none
// ==/UserScript==

(function() {
    'use strict';

    // --- UI 样式 ---
    const containerStyle = `
        position: fixed; bottom: 20px; right: 20px; z-index: 999999;
        display: flex; flex-direction: column; gap: 8px;
        font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif;
    `;

    const btnStyle = `
        padding: 10px 16px; border: none; border-radius: 6px;
        font-size: 13px; font-weight: 600; cursor: pointer; color: white;
        box-shadow: 0 4px 12px rgba(0,0,0,0.15); transition: transform 0.2s, opacity 0.2s;
        display: flex; align-items: center; justify-content: center;
    `;

    // --- 创建 UI ---
    const container = document.createElement('div');
    container.style.cssText = containerStyle;
    document.body.appendChild(container);

    const createBtn = (text, bgColor, id) => {
        const b = document.createElement('button');
        b.innerHTML = text;
        b.id = id;
        b.style.cssText = btnStyle + `background-color: ${bgColor};`;
        b.onmouseenter = () => b.style.transform = "translateY(-2px)";
        b.onmouseleave = () => b.style.transform = "translateY(0)";
        container.appendChild(b);
        return b;
    };

    const btnZip = createBtn("📦 下载图片打包 (ZIP)", "#007bff", "dl-zip");
    const btnPdf = createBtn("📄 下载电子书 (PDF)", "#dc3545", "dl-pdf");

    // 状态提示条
    const statusDiv = document.createElement('div');
    statusDiv.style.cssText = "background: rgba(0,0,0,0.85); color: #fff; padding: 6px 10px; border-radius: 4px; font-size: 12px; display: none; text-align: center; max-width: 200px;";
    container.appendChild(statusDiv);

    const showStatus = (msg, autoHide = false) => {
        statusDiv.style.display = "block";
        statusDiv.innerHTML = msg;
        if (autoHide) setTimeout(() => statusDiv.style.display = "none", 5000);
    };

    // --- 核心依赖加载 ---
    const loadJsPDF = async () => {
        if (window.jspdf) return;
        showStatus("🔧 正在初始化 PDF 组件...");
        return new Promise((resolve) => {
            const script = document.createElement('script');
            script.src = "https://cdnjs.cloudflare.com/ajax/libs/jspdf/2.5.1/jspdf.umd.min.js";
            script.onload = () => { showStatus("✅ 组件就绪", true); resolve(); };
            script.onerror = () => { alert("组件加载失败，请检查网络"); resolve(); }; // 失败也继续，防止卡死
            document.head.appendChild(script);
        });
    };

    // --- 核心逻辑 ---

    // 1. 获取图片列表 (兼容所有变量)
    const getPages = () => {
        const candidates = [
            window.fliphtml5_pages, window.configForPages,
            (window.bookConfig && window.bookConfig.pages),
            (window.htmlConfig && window.htmlConfig.pages),
            (window.sliderConfig && window.sliderConfig.pages),
            (window.articleConfig && window.articleConfig.pages)
        ];
        for (let c of candidates) {
            if (c && Array.isArray(c) && c.length > 0) return c;
        }
        // 暴力扫描 bookConfig
        if (window.bookConfig) {
            for (let key in window.bookConfig) {
                if (Array.isArray(window.bookConfig[key]) && window.bookConfig[key][0]?.path) return window.bookConfig[key];
            }
        }
        return [];
    };

    // 2. 智能探测图片 URL (本次升级核心)
    const probeImage = async (path, baseUrls) => {
        // 彻底清理文件名：去掉开头的 /，去掉 ? 后面的所有参数
        const cleanName = path.split('?')[0].replace(/^\/+/, '');

        // 构造所有可能的路径模板
        // 注意：book.yunzhan365.com 有时候不喜欢 /files/large，而喜欢 /files/mobile
        const patterns = [
            `files/large/${cleanName}`,   // 高清首选
            `files/mobile/${cleanName}`,  // 普通首选
            `files/shot/${cleanName}`,    // 低清备选
            `large/${cleanName}`,         // 变体
            `mobile/${cleanName}`,        // 变体
            `${cleanName}`                // 根目录
        ];

        for (let base of baseUrls) {
            for (let pattern of patterns) {
                // 确保 base 不以 / 结尾，pattern 不以 / 开头，防止双斜杠
                const targetUrl = `${base.replace(/\/$/, '')}/${pattern}`;
                try {
                    // 使用 HEAD 请求快速探测，或者直接 fetch
                    const resp = await fetch(targetUrl);
                    if (resp.ok) return await resp.blob();
                } catch (e) { }
            }
        }
        return null;
    };

    // 3. 计算基准 URL (自动适配 book/bookh)
    const getBaseUrls = () => {
        let currentUrl = window.location.href.split('?')[0];
        // 移除 index.html, mobile 等后缀，拿到书本根目录
        // 例如: https://book.yunzhan365.com/qsrm/tsab/mobile/index.html -> https://book.yunzhan365.com/qsrm/tsab
        let root = currentUrl.replace(/\/mobile\/index\.html$/, '')
                             .replace(/\/mobile\/$/, '')
                             .replace(/\/index\.html$/, '')
                             .replace(/\/$/, '');

        const urls = [root];

        // 自动添加兄弟域名尝试 (book <-> bookh)
        if (root.includes('book.')) urls.push(root.replace('book.', 'bookh.'));
        else if (root.includes('bookh.')) urls.push(root.replace('bookh.', 'book.'));

        return urls;
    };

    // --- 主流程 ---
    const startDownload = async (mode) => {
        const pages = getPages();
        if (!pages.length) return alert("❌ 无法提取图片，请等待书本加载完毕或翻一页再试！");

        if (mode === 'pdf') await loadJsPDF();

        // 锁定 UI
        [btnZip, btnPdf].forEach(b => { b.disabled = true; b.style.opacity = 0.6; });

        const baseUrls = getBaseUrls();
        console.log("探测基准地址:", baseUrls);

        let zip = mode === 'zip' ? new JSZip() : null;
        let pdf = null;
        let successCount = 0;

        for (let i = 0; i < pages.length; i++) {
            showStatus(`⏳ [${i + 1}/${pages.length}] 正在下载...`);

            // 提取文件名
            let p = pages[i];
            let rawPath = (typeof p === 'string') ? p : (p.path || p.url || p.image || (p.n && p.n[0]));
            if (!rawPath) continue;

            // 智能探测下载
            const blob = await probeImage(rawPath, baseUrls);

            if (blob) {
                successCount++;
                if (mode === 'zip') {
                    const ext = blob.type.includes('webp') ? 'webp' : 'jpg';
                    zip.file(`${(i + 1).toString().padStart(3, '0')}.${ext}`, blob);
                } else {
                    // PDF 处理
                    await new Promise((resolve) => {
                        const img = new Image();
                        img.src = URL.createObjectURL(blob);
                        img.onload = () => {
                            const { jsPDF } = window.jspdf;
                            const w = img.naturalWidth;
                            const h = img.naturalHeight;

                            if (!pdf) pdf = new jsPDF({ unit: 'px', format: [w, h], orientation: w > h ? 'l' : 'p' });
                            else pdf.addPage([w, h], w > h ? 'l' : 'p');

                            pdf.addImage(img, blob.type.includes('webp') ? 'WEBP' : 'JPEG', 0, 0, w, h);
                            URL.revokeObjectURL(img.src);
                            resolve();
                        };
                        img.onerror = resolve; // 出错跳过
                    });
                }
            } else {
                console.error(`第 ${i+1} 页下载失败:`, rawPath);
            }
        }

        // 导出文件
        const safeTitle = document.title.replace(/[\\/:*?"<>|]/g, "_") || "云展网电子书";

        if (successCount === 0) {
            alert("❌ 所有图片下载失败！\n可能原因：\n1. 此书开启了特殊加密。\n2. 网络连接问题。");
        } else {
            showStatus("📦 正在打包，请稍候...");
            if (mode === 'zip') {
                const content = await zip.generateAsync({ type: "blob" });
                saveFile(content, `${safeTitle}.zip`);
            } else {
                if (pdf) pdf.save(`${safeTitle}.pdf`);
            }
            showStatus(`🎉 完成！成功下载 ${successCount} 页`, true);
        }

        // 恢复 UI
        [btnZip, btnPdf].forEach(b => { b.disabled = false; b.style.opacity = 1; });
    };

    const saveFile = (blob, filename) => {
        const link = document.createElement("a");
        link.href = URL.createObjectURL(blob);
        link.download = filename;
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
    };

    // 绑定事件
    btnZip.onclick = () => startDownload('zip');
    btnPdf.onclick = () => startDownload('pdf');

})();

File: yunzhan.py
import os
import time
import json
import requests
from concurrent.futures import ThreadPoolExecutor

# 图片处理库 (用于合成 PDF)
from PIL import Image

# 自动化浏览器模块
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager

# ================= 配置区域 =================
MAX_THREADS = 16
# ===========================================

def get_headers():
    return {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36",
    }

def fetch_book_data(url):
    print(f"🕵️‍♂️ 正在启动隐形浏览器 (分析页面)...")
    options = Options()
    
    # 【修改点1】开启无头模式 (隐身)
    options.add_argument("--headless") 
    
    options.add_argument("--disable-gpu")
    options.add_argument("--no-sandbox")
    options.add_argument("--window-size=1920,1080")
    options.add_experimental_option('excludeSwitches', ['enable-logging'])

    driver = None
    try:
        service = Service(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=service, options=options)
        driver.get(url)
        
        print("⏳ 等待加密模块解密 (10 秒)...")
        time.sleep(2) 
        
        # 万能提取脚本
        extract_script = """
        try {
            var candidates = [
                window.fliphtml5_pages, window.configForPages,
                (window.bookConfig && window.bookConfig.pages),
                (window.htmlConfig && window.htmlConfig.pages),
                (window.sliderConfig && window.sliderConfig.pages)
            ];
            var pages = [];
            for(var i=0; i<candidates.length; i++) {
                if(candidates[i] && Array.isArray(candidates[i]) && candidates[i].length > 0) {
                    pages = candidates[i]; break;
                }
            }
            if(pages.length === 0 && window.bookConfig) {
                 for(var key in window.bookConfig) {
                    if(Array.isArray(window.bookConfig[key]) && window.bookConfig[key].length > 0) {
                        var first = window.bookConfig[key][0];
                        if(first && (first.path || first.url || first.image)) {
                            pages = window.bookConfig[key]; break;
                        }
                    }
                }
            }
            var result = [];
            for(var i=0; i<pages.length; i++) {
                var p = pages[i];
                if(typeof p === 'string') result.push(p);
                else if(p.path) result.push(p.path);
                else if(p.url) result.push(p.url);
                else if(p.image) result.push(p.image);
                else if(p.n && p.n[0]) result.push(p.n[0]);
            }
            return { title: document.title, pages: result };
        } catch(e) { return null; }
        """
        return driver.execute_script(extract_script)
    except Exception as e:
        print(f"❌ 浏览器错误: {e}")
        return None
    finally:
        if driver: driver.quit()

def probe_correct_url(base_url, first_path):
    # 去掉文件名中的参数
    clean_name = first_path.split('?')[0].lstrip("/")
    
    print(f"🔍 正在探测路径 (测试文件: {clean_name})...")
    
    base_urls = [base_url]
    if "bookh." in base_url:
        base_urls.append(base_url.replace("bookh.", "book."))

    patterns = [
        "{base}/files/large/{path}",
        "{base}/files/mobile/{path}",
        "{base}/{path}",
        "{base}/large/{path}",
        "{base}/mobile/{path}"
    ]

    for base in base_urls:
        for pattern in patterns:
            test_url = pattern.format(base=base, path=clean_name)
            try:
                r = requests.get(test_url, headers=get_headers(), timeout=5, stream=True, allow_redirects=True)
                if r.status_code == 200:
                    print(f"✅ 路径通了！")
                    return pattern.format(base=base, path="{path}")
            except:
                pass
    return None

def download_image_task(args):
    url_template, filename, save_path, index = args
    try:
        clean_name = filename.split('?')[0].lstrip("/")
        final_url = url_template.format(path=clean_name)
        
        r = requests.get(final_url, headers=get_headers(), timeout=15)
        
        if r.status_code == 200:
            with open(save_path, "wb") as f: f.write(r.content)
            print(f"✅ P{index} OK")
            return
        else:
            print(f"❌ P{index} 失败 ({r.status_code})")
    except Exception as e:
        print(f"❌ P{index} 错误: {e}")

# 【修改点2】新增 PDF 生成函数
def generate_pdf(folder_path, pdf_name):
    print("-" * 30)
    print(f"📑 正在合成 PDF: {pdf_name}")
    
    images = []
    files = [f for f in os.listdir(folder_path) if f.endswith(('.jpg', '.webp', '.png'))]
    # 按文件名排序 (确保 001, 002 顺序正确)
    files.sort()
    
    if not files:
        print("❌ 文件夹为空，无法生成 PDF")
        return

    for f in files:
        try:
            full_path = os.path.join(folder_path, f)
            img = Image.open(full_path)
            # PDF 不支持 RGBA (透明通道)，如果是 WebP 需要转 RGB
            if img.mode != "RGB":
                img = img.convert("RGB")
            images.append(img)
        except Exception as e:
            print(f"⚠️ 跳过损坏图片: {f}")

    if images:
        try:
            output_path = os.path.join(folder_path, pdf_name)
            images[0].save(output_path, "PDF", resolution=100.0, save_all=True, append_images=images[1:])
            print(f"🎉 PDF 生成成功！文件位于: {output_path}")
        except Exception as e:
            print(f"❌ PDF 生成失败: {e}")
    else:
        print("❌ 没有有效的图片用于合成 PDF")

def main():
    print("云展网下载器 v11.0 (全自动隐身 + PDF)")
    print("="*40)
    
    while True:
        url = input("请输入链接 (q 退出):\n>>> ").strip()
        if url.lower() == 'q': break
        if not url: continue
        
        # 1. 解密
        data = fetch_book_data(url)
        if not data or not data.get("pages"):
            print("❌ 解密失败，未获取到页面。")
            continue
            
        pages = data["pages"]
        title = data.get("title", "book")
        print(f"✅ 提取到 {len(pages)} 页。")
        
        # 2. 确定基准 URL
        base_url = url.split("?")[0]
        if "/mobile/" in base_url: base_url = base_url.split("/mobile/")[0]
        elif "/files/" in base_url: base_url = base_url.split("/files/")[0]
        else: base_url = os.path.dirname(base_url)
        
        # 3. 智能探测
        url_template = ""
        if not pages[0].startswith("http"):
            url_template = probe_correct_url(base_url, pages[0])
            if not url_template:
                print("❌ 所有路径尝试均失败。")
                continue
        else:
            url_template = "{path}" 
        
        # 4. 准备下载
        safe_title = "".join([c for c in title if c.isalnum() or c in (' ', '_', '-')]).strip()
        folder = f"{safe_title}_{int(time.time())}"
        os.makedirs(folder, exist_ok=True)
        
        tasks = []
        for i, p in enumerate(pages):
            ext = "webp"
            if ".jpg" in p: ext = "jpg"
            if ".png" in p: ext = "png"
            
            save_path = os.path.join(folder, f"{i+1:03d}.{ext}")
            tasks.append((url_template, p, save_path, i+1))
            
        print(f"🚀 启动下载...")
        with ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            executor.map(download_image_task, tasks)
        
        # 5. 生成 PDF
        generate_pdf(folder, f"{safe_title}.pdf")
        
        print("🎉 全部流程结束！\n")

if __name__ == "__main__":
    main()